Datatype handling when in write_raw_bids

richardkoehler opened the pull request:
PR Description
--------------

Related to #752 

The user must now specify the datatype when calling write_raw_bids. Not automatic datatype inferring is done anymore in write_raw_bids. However, it is checked if the specified datatype is present in the dataset.


Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"

agramfort commented on mne_bids/utils.py:
```diff
@@ -91,20 +91,29 @@ def _get_ch_type_mapping(fro='mne', to='bids'):
+    if 'eeg' in raw:
+        datatypes.append('eeg')
+    if len(datatypes) == 0:
+        raise ValueError('No MEG, EEG or iEEG channels found in data.'
```
```suggestion
        raise ValueError('No MEG, EEG or iEEG channels found in data. '
```

agramfort commented on mne_bids/utils.py:
```diff
@@ -385,3 +394,32 @@ def _stamp_to_dt(utc_stamp):
+    if datatype not in supported_types:
+        raise ValueError(
+            f'The specified datatype {datatype} is currently not supported. '
+            'It should be one of  either `meg`, `eeg` or `ieeg`. Please '
```
```suggestion
            f'It should be one of  either `meg`, `eeg` or `ieeg` (Got `{datatype}`}. Please '
```

agramfort commented on mne_bids/utils.py:
```diff
@@ -385,3 +394,32 @@ def _stamp_to_dt(utc_stamp):
+            'specify a valid datatype using '
+            '`bids_path.update(datatype="<datatype>")`.')
+    datatype_matches = False
+    if datatype == 'eeg':
+        if datatype in raw:
+            datatype_matches = True
```
```suggestion
    if datatype == 'eeg' and datatype in raw:
        datatype_matches = True
```

agramfort commented on mne_bids/utils.py:
```diff
@@ -385,3 +394,32 @@ def _stamp_to_dt(utc_stamp):
+    if datatype == 'eeg':
+        if datatype in raw:
+            datatype_matches = True
+    elif datatype == 'meg':
+        if datatype not in raw:
```
idem

richardkoehler replied:
```suggestion
    elif datatype == 'meg' and datatype in raw:
```

agramfort commented on mne_bids/write.py:
```diff
@@ -1185,7 +1185,11 @@ def write_raw_bids(raw, bids_path, events_data=None,
         raise ValueError(msg)
 
-    datatype = _handle_datatype(raw)
+    if bids_path.datatype is None:
+        datatype = _handle_datatype(raw)
+    else:
+        datatype = bids_path.datatype
+        _check_datatype(raw, datatype)
```
I would do this in the _handle_datatype function

`datatype = _handle_datatype(raw, bids_path.datatype)`

then you call _check_datatype in _handle_datatype and your return directly the 2nd parameter if valid.

richardkoehler replied:
```suggestion
    datatype = _handle_datatype(raw)
```

agramfort reviewed (commented):
just a few suggestions.

thx @richardkoehler

richardkoehler commented:
@agramfort Thanks for the helpful suggestions! Tests are currently failing because write_raw_bids() now makes no more automatic assumptions about the datatype if there are multiple ephys data types (meg, eeg or ieeg) in our Raw object and because it actually checks if the bids_path.datatype is present in the Raw object. I can start fixing the tests, but as I already wrote in #752, I think the current PR might "break" write_raw_bids() for a lot of existing code out there (as demonstrated by the failing tests). So I think we need to make sure that everyone agrees with the new behavior of write_raw_bids:

1. If bids_path.datatype is not set, _handle_datatype will try to infer the datatype:
    a) If there is a single data type (of meg, eeg, ieeg) in the Raw object, this type will be picked
    b) If there are multiple types (of meg, eeg, ieeg) in the Raw object, an Error will be thrown, asking the user to specify the  data type
    c)  If no data type (of meg, eeg, ieeg) is found in the Raw object, an Error will be thrown, asking the user to specify the channel types and/or bids_path.datatype
2. If bids_path.datatype is set, _handle_datatype will call _check_datatype and verify that the specified data type is actually present in the Raw object. If it is not, an Error will be thrown.

Alternatively, we could "trust" the user to know what he is doing, and simply issue a warning message for case 2. There might be some cases where a user might want to write the data without specifying the channel types, even though I can't currently think of any. Is it really the job of MNE-BIDS to enforce these kind of things?
@jasmainak

agramfort commented:
I would say that if you have no datatype and the data contains MEG and EEG
then you should
infer automatically it's an meg data.

would this already avoid breaking many tests?

richardkoehler commented:
@agramfort  Yes, this actually fixed many of the tests. I implemented this way of handling datatype=None and fixed the remaining failing tests. In the case of datatype='eeg' with additional 'meg' channels in the data we are now trusting the user to know what he is doing, and NOT enforcing the datatype 'meg'. Therefore, the user can decide to write out combined MEG-EEG recordings in an EEG data format.

richardkoehler commented:
@agramfort  Actually, it won't be possible as I get the following Error:  
```
E           UserWarning: Encountered unsupported non-voltage units: T/m, T
E           Note that the BrainVision format specification supports only µV.
```
So apparently it doesn't make any sense to save MEG-EEG data as EEG data because brainvision doesn't support MEG units. Sorry, I'm not very familiar with MEG, I wasn't aware of this.

agramfort commented on mne_bids/utils.py:
```diff
@@ -89,22 +89,67 @@ def _get_ch_type_mapping(fro='mne', to='bids'):
+        if 'eeg' in raw:
+            datatypes.append('eeg')
+        if len(datatypes) == 0:
+            raise ValueError('No MEG, EEG or iEEG channels found in data.'
```
```suggestion
            raise ValueError('No MEG, EEG or iEEG channels found in data. '
```

agramfort commented on mne_bids/utils.py:
```diff
@@ -385,3 +430,42 @@ def _stamp_to_dt(utc_stamp):
+    elif datatype == 'meg' and datatype in raw:
+        datatype_matches = True
+    elif datatype == 'ieeg':
+        ieeg_types = ['seeg', 'ecog']
```
```suggestion
        ieeg_types = ('seeg', 'ecog')
```

agramfort commented on mne_bids/utils.py:
```diff
@@ -385,3 +430,42 @@ def _stamp_to_dt(utc_stamp):
+    -------
+    None
+    """
+    supported_types = ['meg', 'eeg', 'ieeg']
```
```suggestion
    supported_types = ('meg', 'eeg', 'ieeg')
```

agramfort commented on mne_bids/tests/test_utils.py:
```diff
@@ -124,3 +144,47 @@ def test_infer_eeg_placement_scheme():
+    raw_meg = mne.io.RawArray(rng.random((3, n_points)) * 1e-6, info_meg)
+    info_ieeg = mne.create_info(['ch1', 'ch2', 'ch3'], sfreq, ['seeg'] * 3)
+    raw_ieeg = mne.io.RawArray(rng.random((3, n_points)) * 1e-6, info_ieeg)
+    datatype = None
+    with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
+                                         'is currently not supported.'):
+        _check_datatype(raw_eeg, datatype)
+    datatype = 'anat'
+    with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
+                                         'is currently not supported.'):
+        _check_datatype(raw_eeg, datatype)
+    datatype = 'eeg'
+    _check_datatype(raw_eeg, datatype)
+    with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
+                                         'was not found in the raw object.'):
+        _check_datatype(raw_meg, datatype)
```
```suggestion
    for datatype, raw in [(None, raw_eeg, ('eeg', raw_meg), ('anat', raw_eeg)]:
        with pytest.raises(ValueError, match=f'The specified datatype {datatype} '
                                             'is currently not supported.'):
            _check_datatype(raw, datatype)
```

can you compress with something like this?

richardkoehler replied:
I did, hope it is fine the way I have solved it now :)

agramfort reviewed (approved):
thx @richardkoehler !

agramfort commented:
thx a lot @richardkoehler !

richardkoehler commented:
@agramfort  Thanks for the merge, happy to contribute :) Stupid question because this is my first contribute to mne-bids: The Merge checklist at the top is not for myself to complete right? Because it says "This is not your own PR". I haven't updated whats_new.rst or anything, but of course I would be happy to do so, if required

agramfort commented:
it's for you to fill to help us review but here I looked myself.

thx again

>

hoechenberger commented:
Thanks @richardkoehler!

hoechenberger commented:
Oh noes, we forgot to add a changelog entry!! @richardkoehler care to add one?

richardkoehler commented:
@hoechenberger Yes, I did so in #779 as I couldn't reopen this PR.
